### PR TITLE
pylexibank 2.0rc1

### PR DESCRIPTION
**Pull request by xrotwang**
`pylexibank` now extends `cldfbench`. Thus, the curation repl is gone, so is the `lexibank` cli. Instead, commands are now invoked as subcommands of `cldfbench`, e.g.
```bash
cldfbench download wold --entry-point lexibank.dataset
```
for commands that are provided by `cldfbench` and
```bash
cldfbench lexibank.makecldf -h
```
for commands that are `pylexibank`-specific.

There's quite a few breaking changes. Here are the ones most likely
to affect individual datasets:
- The `cmd_*` methods have now the signature
  ```python
  def cmd_*(self, args):
  ```
  i.e. accept a single positional argument which is an
  `argparse.Namespace`.
- `Dataset.cmd_install` is now `Dataset.cmd_makecldf`
- `Dataset.raw` is now `Dataset.raw_dir`
- What was previously the `pylexibank.cldf.Dataset` is now
  `pylexibank.cldf.LexibankWriter`. But - as before - a dataset does
  not need to instantiate this class themselves. Instead, it is
  available as `args.writer` in `cmd_makecldf`. As before, this is
  a context manager which
  - wipes `cldf/` clean upon entering, and
  - writes the CLDF data to `cldf/` upon exiting.
- The `wl` attribute of the CLDF writer class - which is a reference
  to the `pycldf.Dataset` object being written - is replaced by
  `pylexibank.cldf.LexibankWriter.cldf`. Note that this is a `property`
  which cannot be accessed outside of a with statement (but inside
  `cmd_makecldf` it's there as `args.writer.cldf`).

### CODE REVIEW

**Comment by LinguList** [comment]
I am having trouble with cldfbench:

```
Traceback (most recent call last):
  File "/home/jmlist/.virtualenvs/base/bin/cldfbench", line 11, in <module>
    load_entry_point('cldfbench', 'console_scripts', 'cldfbench')()
  File "/home/jmlist/.virtualenvs/base/lib/python3.5/site-packages/pkg_resources/__init__.py", line 542, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/jmlist/.virtualenvs/base/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2569, in load_entry_point
    return ep.load()
  File "/home/jmlist/.virtualenvs/base/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2229, in load
    return self.resolve()
  File "/home/jmlist/.virtualenvs/base/lib/python3.5/site-packages/pkg_resources/__init__.py", line 2235, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/jmlist/data/datasets/cldf-datasets/cldfbench/src/cldfbench/__init__.py", line 2, in <module>
    from cldfbench.dataset import *  # noqa: F401, F403
  File "/home/jmlist/data/datasets/cldf-datasets/cldfbench/src/cldfbench/dataset.py", line 12, in <module>
    from cldfbench.cldf import CLDFWriter, CLDFSpec
  File "/home/jmlist/data/datasets/cldf-datasets/cldfbench/src/cldfbench/cldf.py", line 11, in <module>
    from cldfbench.catalogs import Catalog
  File "/home/jmlist/data/datasets/cldf-datasets/cldfbench/src/cldfbench/catalogs.py", line 25, in <module>
    class CachingGlottologAPI(GlottologAPI):
TypeError: str() argument 2 must be str, not tuple
```

I do not understand the error-code here...

**Comment by LinguList** [comment]
Sorry, my pyglottolog is not in syn, I realized...

**Comment by LinguList** [comment]
Just checked this, it works fine here, but I had to fiddle, so making a new virtualenv is probalby required and recommended. 

I just pushed in a v2 branch, what needs to be adjusted in chen-hmongmien, and it is not that much, but whith the 30 datasets I have as a patron now, I'll need to devote an afternoon to get this right. 

I'll gladly do this, as it seems worthwhile, I also spotted lots of empty imports when just only checking the datasets installed on my machine, which all imported metadata, although they don't need it. 

So basically, I can nicely live with that, and I suggest the next stage needs to be every lexibanker to start working on lifting hte datasets to the next level.

**Comment by LinguList** [comment]
[here](https://github.com/lexibank/chenhmongmien/compare/master...v2?expand=1#diff-870a3af6d592fadab1b1d7c24655cfb6) is the example of chenhmong mien, what it will need for each dataset more or less, to be modified.

**Comment by xrotwang** [comment]
Regarding configuration: I thought this was introducing more intransparency. But considering that with `cldfcatalog` we can checkout releases "on-the-fly" for just one operation, keeping long-lived clones of the catalogs makes even more sense - so having these paths in a config file would be useful.

I wouldn't go to the lengths of prompting users for automated creation of such a config file - or at most, have a `cldfbench` command to do so.

**Comment by xrotwang** [comment]
See https://github.com/cldf/cldfbench/issues/7

**Comment by LinguList** [comment]
So I think, when all have agreed, everybody, i.e., @tresoldi, @chrzyki, @SimonGreenhill, and @xrotwang and me should split up, who looks at what dataset (we have the google docs for this), and make sure they also do a good checking of unnecessary code by then.

**Comment by xrotwang** [comment]
I think we should 
- experiment a bit longer, with a couple of datasets
- then actually make a release of `pylexibank` 2.0rc1
- then continue with more datasets, which could then properly install pylexibank from pypi

**Comment by LinguList** [comment]
Okay, this would then mean that every lexibanker should, say, look at 2 of their datasets to see how it works?

**Comment by LinguList** [comment]
I guess, most of the changes is indisiputable, e.g., the namespace of models instead of pulling Concept from dataset will avoid errors, etc.

**Comment by xrotwang** [comment]
@LinguList not yet released, but if you install `cldfbench` from clone, you can now get config file support - for just the catalog paths.

**Comment by SimonGreenhill** [comment]
I'll look over this next week. It'll take me some time to get my head around.

However, if we're going to break things now, can we break other things too and get rid of as much boilerplate? e.g. 

1. can we simplify the SplittingDataset vs. NonSplittingDataset thing? Does it make more sense to have Non- as the base form? Or can we just make everything NonSplitting (=Dataset) by having  `split_forms` defined as a no-op, and splitting datasets need to defined `split_forms` themselves?

2. what about progress bars? we seem to have some datasets with tqdm, some with pb, and some with nothing. Can we standardise or drop them? 

3. We have empty calls to `ds.add_languages()`, `ds.add_sources()`, `ds.add_concepts()`. Does it make sense to do these implicitly (in super(ds)), so they don't need to be in the dataset code? On the one hand it's nice to have them explicitly, but on the other they're just boilerplate that could be cleaned out.

**Comment by xrotwang** [comment]
ad 3) except for `add_sources`, I don't think these calls are boilerplate, because they make assumptions about what's in `etc/languages|concepts.csv` which are not enforced by `pylexibank`. Thus, they can't be called "unsupervised". Although, we could go for enforcing "`etc/languages.csv` must contain rows that can be paased to `add_languages`", but I think that's not a good trade-off.

**Comment by LinguList** [comment]
> can we simplify the SplittingDataset vs. NonSplittingDataset thing? Does it make more sense to have Non- as the base form? Or can we just make everything NonSplitting (=Dataset) by having split_forms defined as a no-op, and splitting datasets need to defined split_forms themselves?

Yes, we can get rid of it now, as splitting is explicit. I am well-set with v2 now, so I can make a proposal. I'd probably even add a simple add_lexeme command as a default that does never split, and never adds forms. Ideally, the decision is the function, not the dataset class, and we could even go as far as saying: the default is that you don't split, to avoid abuse (we had some heavily duplicated entries in some datasets, I worked on).

**Comment by LinguList** [comment]
> what about progress bars? we seem to have some datasets with tqdm, some with pb, and some with nothing. Can we standardise or drop them?

Standardize, as it is essential for developing (=adding a new dataset), in my opinion. I am now thinking: is it feasible to think of something like pep8 for lexibank datasets, i.e., a script running through code and pointing to:

1. unused imports (we have a lot)
2. unconventional treatment
3. overridden functions

or is that too much?

**Comment by LinguList** [comment]
as to point 3): we have standardized functions here, ideally where `etc/languages.csv` matches the cldf. I started adjusting all of my patron-datasets to that style, also for concepts, but this is a code-style issue, and maybe relates to my idea to have a simple pep8-style code checker, as it would make explicit that one does not use the default (which would be `add_languages()` as opposed to `add_languages(**params)`

File: setup.py [hunk 1]
```diff
@@ -17,22 +17,26 @@
...
         'attrs>=18.1.0',
         'pyglottolog>=2.0',
         'pyconcepticon>=2.1.0',
         'pyclts>=1.2.0',
         'segments>=2.0.2',
-        'lingpy>=2.6.4',
+        'lingpy>=2.6.5',
         'appdirs',
         'requests',
         'termcolor',
```
[the remark this answers is not in the document]
Comment (SimonGreenhill) [reply setup.py]: do we need segments as well? (isn't this pulled in via clts?)

File: src/pylexibank/commands/ls.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+"""
+List lexibank datasets - installed or loaded into the db.
+"""
+import collections
+import textwrap
+
+import termcolor
```
Comment (SimonGreenhill) [inline src/pylexibank/commands/ls.py]: ahh, there's termcolor.

File: src/pylexibank/db.py [hunk 3]
```diff
@@ -77,9 +74,9 @@ def insert(db, table, keys, *rows, **kw):
             keys = [k.strip() for k in keys.split(',')]
         sql = "INSERT INTO {0} ({1}) VALUES ({2})".format(
             table, ','.join(keys), ','.join(['?' for _ in keys]))
-        if kw.get('verbose'):  # pragma: no cover
-            print(sql)
-            print(rows)
+        #if kw.get('verbose'):  # pragma: no cover
```
Comment (SimonGreenhill) [inline src/pylexibank/db.py]: wouldn't this be better logged to logging.debug?

File: src/pylexibank/report.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+def build_status_badge(dataset):
+    if dataset.dir.joinpath('.travis.yml').exists() and dataset.repo and dataset.repo.github_repo:
+        return "[![Build Status](https://travis-ci.org/{0}.svg?branch=master)]" \
+               "(https://travis-ci.org/{0})".format(dataset.repo.github_repo)
+    return ''
+
+
+def report(dataset, tr_analysis=None, glottolog=None, log=None):
+    #
+    # FIXME: write only summary into README.md
```
Comment (SimonGreenhill) [inline src/pylexibank/report.py]: does this need to actually be fixed? I'm ok with long README files..

File: src/pylexibank/report.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    badges = nfilter([build_status_badge(dataset)])
+
+    for cldf_spec in dataset.cldf_specs_dict.values():
+        lines.extend(cldf_report(cldf_spec, tr_analysis, badges, log, glottolog))
+        break
+    return '\n'.join(lines)
+
+
+def cldf_report(cldf_spec, tr_analysis, badges, log, glottolog):
```
Comment (SimonGreenhill) [inline src/pylexibank/report.py]: Cool, much cleaner, than the mess we had before. Can we expose this as a CLI command too?

File: tests/repos/datasets/test_dataset_cldf/tdc.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from pathlib import Path
```
Comment (SimonGreenhill) [inline tests/repos/datasets/test_dataset_cldf/tdc.py]: td.py and tdc.py are a bit opaque?

File: tests/test_forms.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+
+def test_split():
+    spec = FormSpec()
+    assert spec.split({}, 'x', lexemes={'x': 'x(a);y'}) == ['x', 'y']
+    assert spec.split({}, 'x', lexemes={'x': 'x;?'}) == ['x']
+
+    spec.strip_inside_brackets = False
+    assert spec.split({}, 'x(a)') == ['x(a)']
+
```
Comment (SimonGreenhill) [inline tests/test_forms.py]: is it worth checking a more complex form here too e.g.:

```python
assert spec.split({}, 'x', lexemes={'x': 'x(a);y'}) == ['x', 'x(a)', 'y']
```

File: tests/test_lingpy_util.py [hunk 8]
```diff
@@ -1,71 +1,60 @@
...
+    dataset.language_class = pbds.Language
+    dataset.concept_class = pbds.Concept
+    dataset.split_forms = lambda _, s: [s]
+    dataset.dir = dsdir
+    dataset.tr_analyses = {}
+    dataset.cldf_dir = dsdir.joinpath('cldf')
+
+    dataset.lexeme_class = Lexeme
+    with dataset.cldf_writer(mocker.Mock()) as ds:
+        # needs to be fixed XXX
```
Comment (SimonGreenhill) [inline tests/test_lingpy_util.py]: TODO?

File: setup.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (xrotwang) [reply setup.py]: No, `segments` is still used for the tokenizer.

File: src/pylexibank/report.py (hunk 4 above)
Comment (xrotwang) [reply src/pylexibank/report.py]: I don't even recollect whether it was me adding this note ...

File: src/pylexibank/report.py (hunk 5 above)
Comment (xrotwang) [reply src/pylexibank/report.py]: We could - but at some point we decided to roll it into `_cmd_makecldf`, presumably because this will re-use initialized Glottolog and Concepticon?

**Comment by chrzyki** [comment]
> > what about progress bars? we seem to have some datasets with tqdm, some with pb, and some with nothing. Can we standardise or drop them?
> 
> Standardize, as it is essential for developing (=adding a new dataset), in my opinion. I am now thinking: is it feasible to think of something like pep8 for lexibank datasets, i.e., a script running through code and pointing to:
> 
>     1. unused imports (we have a lot)
> 
>     2. unconventional treatment
> 
>     3. overridden functions
> 
> 
> or is that too much?

A majority of this could be handled with PyCharm (or similar) & format code & optimize import options in conjunction with tools like https://github.com/psf/black. Maintaining a dedicated tool for this seems a bit cumbersome.

File: src/pylexibank/report.py (hunk 5 above)
Comment (SimonGreenhill) [reply src/pylexibank/report.py]: Hmm ok. Is it useful to be able to run the report separately too?

**Comment by LinguList** [comment]
Hi, @SimonGreenhill, @xrotwang, @chrzyki and @tresoldi, I was just thinking: testing the new lexibank might be even done best by adding new datasets, instead of adding old ones? The reason is: adding new ones would be even more benefitial, to test how this works from scratch, e.g., also with some [template](https://github.com/lexibank/template). 

Here's a collection of straightforward data points:

1. [Master thesis by Joo Lan](https://github.com/lexibank/lexibank/issues/177) (we have scheduled a meeting, he can easily provide, e.g., concepticon mappings, and language mappings, I am convinced)
2. [Significance testing Altaic](https://github.com/lexibank/lexibank/issues/176): one needs to edit th PDF via adobe-pro or similar, but from then on, it should be easy to prepare the data, but it does not offer IPA, and no forms.
3. [South American Pronouns](https://github.com/lexibank/lexibank/issues/174): should be very straightforward, as there's a github repo with the data in lingpy-compatible format
4. [Lexical data in Gilman 2012](https://github.com/lexibank/lexibank/issues/175): very easy, I have the data 

And we have the reflex data as well (Polyglotta Africana). So maybe, we could, instead of only adjusting existing lexibank datasets, have each of us a try on doing one from scratch and asking for the others to review the code?

**Comment by LinguList** [comment]
> A majority of this could be handled with PyCharm (or similar) & format code & optimize import options in conjunction with tools like https://github.com/psf/black. Maintaining a dedicated tool for this seems a bit cumbersome.

Regarding this point: I agree that a tool is probably not needed, but ideally, one would have a checklist, and given how many lines of unused code we have in all existing datasets, it shows that we need to find a way of systematically looking into this, I'd say. Ideally, some off-the-shelf tool that does a basic format/code check would be useful, but I am not experienced with this. Yet, running pep8 may already be a start.

File: src/pylexibank/dataset.py [hunk 9]
```diff
@@ -207,6 +52,7 @@ def stats(self):
         return {}
 
     def __init__(self, concepticon=None, glottolog=None):
+        super().__init__()
         if self.__class__ != Dataset:
             if not self.dir:
```
[the remark this answers is not in the document]
Comment (xrotwang) [reply src/pylexibank/dataset.py]: I think that was a bit of hack for testing. Dunno if it is even required anymore.

**Comment by xrotwang** [comment]
Guess we are past the point of no return - so off to 2.0 - I'm merging!
